overlord/hookstate/ctlcmd/helpers.go

```diff
@@ -69,20 +69,32 @@ func getServiceInfos(st *state.State, snapName string, serviceNames []string) ([
 
 var servicestateControl = servicestate.Control
 
-func queueCommand(context *hookstate.Context, ts *state.TaskSet) {
-	// queue command task after all existing tasks of the hook's change
+func queueCommand(context *hookstate.Context, ts *state.TaskSet) error {
+	hookTask, ok := context.Task()
+	if !ok {
+		return fmt.Errorf("attempted to queue command with ephemeral context")
+	}
+
 	st := context.State()
 	st.Lock()
 	defer st.Unlock()
 
-	task, ok := context.Task()
-	if !ok {
-		panic("attempted to queue command with ephemeral context")
+	change := hookTask.Change()
+	hookTaskLanes := hookTask.Lanes()
+	tasks := change.LaneTasks(hookTaskLanes...)
+	if len(hookTaskLanes) == 1 && hookTaskLanes[0] == 0 {
+		hookTaskLanes = nil
+	}
+	for _, l := range hookTaskLanes {
+		ts.JoinLane(l)
 	}
-	change := task.Change()
-	tasks := change.Tasks()
 	ts.WaitAll(state.NewTaskSet(tasks...))
 	change.AddAll(ts)
+	// As this can be run from what was originally the last task of a change,
+	// make sure the tasks added to the change are considered immediately.
+	st.EnsureBefore(0)
```

Comment:
I'm not sure this is something that belongs in queueCommand, it feels like unexpected side-effect.

Reply:
The is needed because we are adding task(s) after the last task that has already been processed. Not having this causes a very long pause at runtime. We do similiar thing in a couple of different places. I'm not sure if there is a better place to do this.

Reply:
If @pedronis agrees then +1